<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span><ul class="toc-item"><li><span><a href="#Style" data-toc-modified-id="Style-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Style</a></span></li><li><span><a href="#Imports-and-Data" data-toc-modified-id="Imports-and-Data-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Imports and Data</a></span></li></ul></li><li><span><a href="#Prices" data-toc-modified-id="Prices-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Prices</a></span><ul class="toc-item"><li><span><a href="#Clean-for-new-sql" data-toc-modified-id="Clean-for-new-sql-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Clean for new sql</a></span><ul class="toc-item"><li><span><a href="#Reindex" data-toc-modified-id="Reindex-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Reindex</a></span></li></ul></li><li><span><a href="#Post-to-sql" data-toc-modified-id="Post-to-sql-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Post to sql</a></span></li><li><span><a href="#Pickle" data-toc-modified-id="Pickle-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Pickle</a></span></li></ul></li><li><span><a href="#Splits" data-toc-modified-id="Splits-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Splits</a></span></li><li><span><a href="#Performance" data-toc-modified-id="Performance-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Performance</a></span><ul class="toc-item"><li><span><a href="#Functions" data-toc-modified-id="Functions-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Functions</a></span></li><li><span><a href="#Clean-for-new-sql" data-toc-modified-id="Clean-for-new-sql-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Clean for new sql</a></span><ul class="toc-item"><li><span><a href="#Analyze-null-values" data-toc-modified-id="Analyze-null-values-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Analyze null values</a></span></li><li><span><a href="#More-cleaning" data-toc-modified-id="More-cleaning-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>More cleaning</a></span></li><li><span><a href="#Execute-col_to_fill" data-toc-modified-id="Execute-col_to_fill-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Execute col_to_fill</a></span></li><li><span><a href="#Execute-col_to_fill_then_drop" data-toc-modified-id="Execute-col_to_fill_then_drop-4.2.4"><span class="toc-item-num">4.2.4&nbsp;&nbsp;</span>Execute col_to_fill_then_drop</a></span></li><li><span><a href="#Add-price-to-performance" data-toc-modified-id="Add-price-to-performance-4.2.5"><span class="toc-item-num">4.2.5&nbsp;&nbsp;</span>Add price to performance</a></span></li><li><span><a href="#Apply-Splits" data-toc-modified-id="Apply-Splits-4.2.6"><span class="toc-item-num">4.2.6&nbsp;&nbsp;</span>Apply Splits</a></span></li><li><span><a href="#Separate-out-penny-stocks" data-toc-modified-id="Separate-out-penny-stocks-4.2.7"><span class="toc-item-num">4.2.7&nbsp;&nbsp;</span>Separate out penny stocks</a></span></li><li><span><a href="#View-loss" data-toc-modified-id="View-loss-4.2.8"><span class="toc-item-num">4.2.8&nbsp;&nbsp;</span>View loss</a></span></li></ul></li><li><span><a href="#Pickle" data-toc-modified-id="Pickle-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Pickle</a></span></li><li><span><a href="#Post-to-sql" data-toc-modified-id="Post-to-sql-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Post to sql</a></span></li></ul></li><li><span><a href="#Company" data-toc-modified-id="Company-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Company</a></span><ul class="toc-item"><li><span><a href="#Clean-for-new-sql" data-toc-modified-id="Clean-for-new-sql-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Clean for new sql</a></span></li></ul></li><li><span><a href="#Analyst" data-toc-modified-id="Analyst-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Analyst</a></span><ul class="toc-item"><li><span><a href="#Checking-for-null-values" data-toc-modified-id="Checking-for-null-values-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Checking for null values</a></span></li><li><span><a href="#Filling-null-values" data-toc-modified-id="Filling-null-values-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Filling null values</a></span></li><li><span><a href="#Mapping" data-toc-modified-id="Mapping-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Mapping</a></span></li><li><span><a href="#Convert-strings-to-float" data-toc-modified-id="Convert-strings-to-float-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Convert strings to float</a></span></li><li><span><a href="#Add-to-sql-server" data-toc-modified-id="Add-to-sql-server-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Add to sql server</a></span></li><li><span><a href="#Pickle" data-toc-modified-id="Pickle-6.6"><span class="toc-item-num">6.6&nbsp;&nbsp;</span>Pickle</a></span></li></ul></li><li><span><a href="#Combine-company,-analyst-and-performance" data-toc-modified-id="Combine-company,-analyst-and-performance-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Combine company, analyst and performance</a></span><ul class="toc-item"><li><span><a href="#Performance" data-toc-modified-id="Performance-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Performance</a></span></li><li><span><a href="#Penny" data-toc-modified-id="Penny-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Penny</a></span></li><li><span><a href="#Write-pickles" data-toc-modified-id="Write-pickles-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Write pickles</a></span></li></ul></li><li><span><a href="#Post-to-firebase" data-toc-modified-id="Post-to-firebase-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Post to firebase</a></span></li></ul></div>

# Introduction
> To do:
> * update firebase vs overwrite
> * take index from performance and penny performance rather than searching through all 8000 stocks

## Style

In [1]:
import warnings
import matplotlib as mpl
mpl.rcParams['axes.titlesize']='xx-large'
mpl.rcParams['xtick.labelsize'] = 13
mpl.rcParams['ytick.labelsize'] = 13
mpl.rcParams['axes.labelsize'] = 16

# Styling notebook
from IPython.core.display import HTML
def css_styling():
    styles = open("./styles/custom.css", "r").read()
    return HTML(styles)
css_styling()

## Imports and Data

In [2]:
USE_GPU = True
IGNORE_WARN = True
SEED = 42

import os
import sys
import time
import datetime
import importlib
import pandas as pd
import matplotlib.pyplot as plt
pd.set_option('display.float_format', lambda x: '%.3f' % x)

# Imports for reproducability
import numpy as np
np.random.seed(SEED)

import tensorflow as tf
tf.random.set_seed(SEED)

if 'mlframe' in sys.modules:
    importlib.reload(sys.modules['mlframe'])
    importlib.reload(sys.modules['db'])

from db import DataBase
from mlframe import MLFrame

# Ignore warnings
if IGNORE_WARN:
    warnings.filterwarnings('ignore')

For convinent loading 'from mlframe import MLFrame


C:\Users\12147\anaconda3\envs\tf-gpu\lib\site-packages\sklearn\utils\deprecation.py:143: FutureWarning: The sklearn.metrics.classification module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.metrics. Anything that cannot be imported from sklearn.metrics is now part of the private API.
  warnings.warn(message, FutureWarning)


In [3]:
db = DataBase('stock', 14333, 'localhost')
db.cur.execute(
"""
SHOW TABLES
""")
db_names = list(table[0] for table in db.cur.fetchall())
db.pull_names_as_dataframes(db_names)
db.conn.close()

# Prices

## Clean for new sql

In [4]:
# Cleaning first blank row
df_prices = db.frames['prices'].copy(deep=True).drop(0)

# Resetting date index and dropping index
df_prices.set_index('date', inplace=True, drop=True)

print("Null values = \n", df_prices.isna().sum())
print("\n\nIndexes")
query = df_prices.query("sym == 'AAPL'")
aapl_dis = query.loc[query.index > pd.to_datetime('2019-08-08')].reset_index()
display(aapl_dis.head(1))
display(aapl_dis.tail(1))
print("\n\n Head")
df_prices.head()

Null values = 
 sym      0
price    0
dtype: int64


Indexes


date   sym       price
0  2019-08-09  AAPL  200.940000

date   sym       price
338  2020-12-01  AAPL  122.720000



 Head


sym      price
date                     
2019-06-04   A  67.950000
2019-06-08   A  69.520000
2019-06-17   A  70.290000
2019-06-19   A  71.910000
2019-06-20   A  73.320000

### Reindex

In [5]:
idx = pd.date_range(aapl_dis.head(1).date.values[0], aapl_dis.tail(1).date.values[0], freq='B')
drop_index = pd.to_datetime(['2019-12-09', '2020-06-23', '2020-06-24', '2020-06-25'])
idx = idx.drop(drop_index)

all_symbols = df_prices['sym'].unique()

sym_dfs = []
unfilled_syms = []
sym_num = len(all_symbols)
times = np.array([])

def cust_interpolate_fill_na(sym):
    df = pd.DataFrame(df_prices.query(f"sym == '{sym}'"), index=idx)
    df['price'] = pd.to_numeric(df['price'], errors='coerce').interpolate()
    df['sym'].fillna(sym, inplace=True)
    if df.isna().sum().sum():
        unfilled_syms.append(sym)
    else:
        sym_dfs.append(df)

for num, sym in enumerate(all_symbols):
    start = time.time()
    cust_interpolate_fill_na(sym)  # Timed function
    end = time.time()
    time_taken = end-start
    times = np.append(times, time_taken)
    print(f"{num+1}/{sym_num} in {time_taken: .2f}s for {sym} :: averaging {times.mean():.2f}s/sym...", end='\r')

df_prices = pd.concat(sym_dfs)
df_prices.index = df_prices.index.strftime('%Y-%m-%d')
df_prices.index.name = 'date'
df_prices.head()

sym  price
date                 
2019-08-09   A 68.640
2019-08-12   A 66.780
2019-08-13   A 67.640
2019-08-14   A 65.590
2019-08-15   A 69.590

In [6]:
print("Total stocks left = ", df_prices['sym'].nunique())
print("NA values = ", df_prices.isna().sum().sum())

Total stocks left =  6103
NA values =  0


## Post to sql

In [7]:
# Reindex for entry
df_prices = df_prices.reset_index().set_index(['sym', 'date'], drop=True)

In [8]:
# Open database cleaned
db_clean = DataBase('stock_cleaned', 14333, 'localhost')
db_clean.pull_names_as_dataframes(['prices'])
db_clean.frames['prices']

# Make list of queries to post to sql
query_list = []
for idx, row in df_prices.reset_index().iterrows():
    query_list.append(tuple(row.to_dict().values()))

# Execute query
db_clean.cur.executemany("""
REPLACE INTO prices (sym, date, price)
VALUES (%s, %s, %s)""",
      query_list)

# Commit
db_clean.conn.commit()

# Close db
db_clean.conn.close()

## Pickle

In [9]:
# Pickle because of 30 minute runtime.  Need to find a faster way...
df_prices.to_pickle("./data/prices.pkl")

# Assert pickle
assert (len(df_prices) == (df_prices == pd.read_pickle("./data/prices.pkl")).sum()).values[0]

In [10]:
df_prices = pd.read_pickle("./data/prices.pkl")
df_prices = df_prices.reset_index().set_index(['date'])
df_prices.index = pd.to_datetime(df_prices.index)

# Splits

In [11]:
# Load in from database frame
df_splits = db.frames['splits']

# Clear dates older than beginning index of prices
df_splits = df_splits[df_splits['execute'] > pd.to_datetime('2019-08-08')]

# Get num/den as floats to make an apply column which will be used
# when applying the actual split
df_splits['num'] = df_splits['num'].astype('float32')
df_splits['den'] = df_splits['den'].astype('float32')
df_splits['apply'] = df_splits['num'] / df_splits['den']

# Display
print("Total splits =", len(df_splits))
print("Total split p/stock =", df_splits['sym'].value_counts().values)
df_splits.head()

Total splits = 254
Total split p/stock = [7 3 2 2 2 2 2 2 2 2 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]


sym   num       den    announce      record     execute  apply
0  AAPL 4.000     1.000  2020-07-30  2020-08-24  2020-08-31  4.000
1  ACHV 1.000    20.000  2020-07-30  2020-07-31  2020-07-31  0.050
2  AEMD 1.000    15.000  2019-10-14  2019-10-15  2019-10-15  0.067
3   AHT 1.000    10.000  2020-07-02  2020-07-16  2020-07-16  0.100
6  AITX 1.000 10000.000  2020-03-26  2020-03-27  2020-03-27  0.000

# Performance
> **Steps to take**
> * Load in performance, and clean
> * Match index to prices
> * Try to interpolate null values by one day as we did with prices
> * Fill null ExDividend dates with 1970-01-01 then encode days since then for numerical data
> * Decide what columns to drop, and which stocks to drop if they are in a given column
> * Apply splits
> * Post to sql server  CHECKPOINT
> * Drop penny stocks ~1/5 of data

## Functions

In [12]:
def get_perc_not_null(df):
    """Returns the percentages of null in each column to the length of the df"""
    nulls = df.isna().sum()
    return pd.Series(1 - nulls/len(df)).sort_values() * 100

In [13]:
def check_quantity_stocks_null_in_columns(df, cols):
    """Query to find how many stocks have a null value in any of these columns
        loc the rows that have any nulls in the given columns
        reset the index to 0-x
        get 'sym' series
        get the number of unique values
        """
    return df.loc[df[cols].isna().transpose().any()].reset_index()['sym'].nunique()

In [14]:
def try_replace(x):
    """Quick fix to some of the data having '1,300' instead of a number"""
    try:
        x = x.replace(',', '')
        return x
    except Exception:
        return x

## Clean for new sql

In [15]:
# These are some dropped columns that were obtained after executing the below
# We now drop them before to reduce the processing time
drop_cols = ['PEG', 'FwdEPS3to5Yr', 'DividendCoverage', 'TotalReturn10Yr',
             'EPSGrowth5Year', 'TotalReturn5Yr']

In [16]:
# Cleaning first two blank rows
df_performance = db.frames['performance'].copy(deep=True).drop([0, 1])

# Drop cols
df_performance.drop(drop_cols, inplace=True, axis=1)

# Remove Dates that are less than august 8th, as that is from a different collection method
df_performance = df_performance[df_performance['date'] > pd.to_datetime('2019-08-08')]

# Resetting date index and dropping index
df_performance.set_index('date', inplace=True, drop=True)

# Dropping null date values
df_performance = df_performance.loc[df_performance.index.notnull()]

# Replace -- with NaN
df_performance.replace(['', '--'], np.NaN, inplace=True)
df_performance.head()

sym         MCAP RevenueTTM TotalCash ExDividendDate Dividend  \
date                                                                       
2019-08-09   A  22027.70459       5019      2155            NaN    0.164   
2019-08-12   A  21758.26224       5019      2155            NaN    0.164   
2019-08-13   A  21121.11033       5019      2155            NaN    0.164   
2019-08-14   A  21441.27124       5019      2155            NaN    0.164   
2019-08-15   A      20332.9       5090      1765            NaN    0.164   

           DividendYield PriceToBook PriceToSales EnterpriseValue  ...    PE  \
date                                                               ...         
2019-08-09       0.94402     4.29809      4.40182     21670704590  ...  19.8   
2019-08-12       0.95571     4.24551      4.34797     21401262240  ...  19.2   
2019-08-13       0.98454     4.12119      4.22065     20764110330  ...  19.5   
2019-08-14       0.96984     4.18366      4.28463     21084271240  ...  18.9   
2019-08-15       1.00015     4.28332      4.06782     20365900000  ...  20.8   

           ReturnonEquity FreeCashFlow TotalDebtToEquity  \
date                                                       
2019-08-09        23.1598          887          37.29007   
2019-08-12        23.1598          887          37.29007   
2019-08-13        23.1598          887          37.29007   
2019-08-14        23.1598          887          37.29007   
2019-08-15        22.0179          918          36.93453   

           PricePerformance52Weeks PricePerformanceYTD TotalReturn1Yr  \
date                                                                    
2019-08-09                 3.13149             3.00919        4.04369   
2019-08-12                 3.59191             1.74918        4.50818   
2019-08-13                 0.55841            -1.23036        1.44784   
2019-08-14                  2.5781             0.26682         3.4854   
2019-08-15                -1.73783            -2.77201        -0.8687   

           TotalReturn3Yr SharesOutstanding VolumeAvg90Day  
date                                                        
2019-08-09       14.79356           315.993        2.59274  
2019-08-12       14.09856           315.993        2.58484  
2019-08-13        12.4788           315.993        2.60485  
2019-08-14       13.04428           315.993        2.61616  
2019-08-15       11.89052           315.993        2.63078  

[5 rows x 21 columns]

In [17]:
# Get all symbols from price and split out only the symbols that are in price
df_prices = df_prices.reset_index().set_index(['date'])
prices_symbols = df_prices['sym'].unique()
df_performance = df_performance.loc[df_performance['sym'].isin(prices_symbols)]
df_prices.head()

sym  price
date                 
2019-08-09   A 68.640
2019-08-12   A 66.780
2019-08-13   A 67.640
2019-08-14   A 65.590
2019-08-15   A 69.590

#### Add prices to performance

In [18]:
test_performance = df_performance.reset_index().set_index(['date', 'sym'])
test_prices = df_prices.reset_index().set_index(['date', 'sym'])
df_performance = pd.DataFrame(test_performance, index=test_prices.index)
assert (df_performance.shape[0] == df_prices.shape[0])

df_performance.head()

MCAP RevenueTTM TotalCash ExDividendDate Dividend  \
date       sym                                                             
2019-08-09 A    22027.70459       5019      2155            NaN    0.164   
2019-08-12 A    21758.26224       5019      2155            NaN    0.164   
2019-08-13 A    21121.11033       5019      2155            NaN    0.164   
2019-08-14 A    21441.27124       5019      2155            NaN    0.164   
2019-08-15 A        20332.9       5090      1765            NaN    0.164   

               DividendYield PriceToBook PriceToSales EnterpriseValue  \
date       sym                                                          
2019-08-09 A         0.94402     4.29809      4.40182     21670704590   
2019-08-12 A         0.95571     4.24551      4.34797     21401262240   
2019-08-13 A         0.98454     4.12119      4.22065     20764110330   
2019-08-14 A         0.96984     4.18366      4.28463     21084271240   
2019-08-15 A         1.00015     4.28332      4.06782     20365900000   

               PriceToCashFlow    PE ReturnonEquity FreeCashFlow  \
date       sym                                                     
2019-08-09 A          16.57368  19.8        23.1598          887   
2019-08-12 A          16.37095  19.2        23.1598          887   
2019-08-13 A          15.89156  19.5        23.1598          887   
2019-08-14 A          16.13245  18.9        23.1598          887   
2019-08-15 A          16.05056  20.8        22.0179          918   

               TotalDebtToEquity PricePerformance52Weeks PricePerformanceYTD  \
date       sym                                                                 
2019-08-09 A            37.29007                 3.13149             3.00919   
2019-08-12 A            37.29007                 3.59191             1.74918   
2019-08-13 A            37.29007                 0.55841            -1.23036   
2019-08-14 A            37.29007                  2.5781             0.26682   
2019-08-15 A            36.93453                -1.73783            -2.77201   

               TotalReturn1Yr TotalReturn3Yr SharesOutstanding VolumeAvg90Day  
date       sym                                                                 
2019-08-09 A          4.04369       14.79356           315.993        2.59274  
2019-08-12 A          4.50818       14.09856           315.993        2.58484  
2019-08-13 A          1.44784        12.4788           315.993        2.60485  
2019-08-14 A           3.4854       13.04428           315.993        2.61616  
2019-08-15 A          -0.8687       11.89052           315.993        2.63078

### Analyze null values

In [19]:
get_perc_not_null(df_performance)

ExDividendDate             7.486
DividendYield             34.444
Dividend                  34.636
PE                        45.338
PriceToCashFlow           53.868
PriceToSales              75.244
FreeCashFlow              75.711
TotalReturn3Yr            77.774
PriceToBook               78.177
RevenueTTM                78.602
EnterpriseValue           79.158
TotalDebtToEquity         79.572
ReturnonEquity            83.120
PricePerformanceYTD       84.625
TotalCash                 85.406
MCAP                      86.419
VolumeAvg90Day            86.453
TotalReturn1Yr            88.693
PricePerformance52Weeks   88.693
SharesOutstanding         92.520
dtype: float64

#### Convert all values to float

In [20]:
performance_mins = {}
for col in df_performance.columns:
    try:
        performance_mins[col] = df_performance[col].map(try_replace).fillna(0).astype('float32').min()
    except Exception:
        print(col, 'Exception ---')
performance_mins = pd.Series(performance_mins)
performance_mins.sort_values()

ExDividendDate Exception ---


EnterpriseValue           -278943858688.000
ReturnonEquity                 -1255335.250
TotalDebtToEquity               -288322.562
FreeCashFlow                     -84924.000
RevenueTTM                         -811.925
TotalReturn1Yr                      -99.998
PricePerformance52Weeks             -99.998
PricePerformanceYTD                 -99.993
TotalReturn3Yr                      -99.767
MCAP                                  0.000
PE                                    0.000
PriceToCashFlow                       0.000
PriceToSales                          0.000
PriceToBook                           0.000
DividendYield                         0.000
Dividend                              0.000
TotalCash                             0.000
SharesOutstanding                     0.000
VolumeAvg90Day                        0.000
dtype: float64

<div class="alert alert-warning shadow">
  <strong>Filling Null Values:</strong>
    <ul>
    <li><b>ExDividendDate:</b> Not all of the dividend stocks have a dividend date, so we will be filling null values of this column with the time equivalent of 0 1970-01-01</li></ul>
        <br></br>
    <b>For the rest of the stocks we will be doing two different things:</b>
    <ul>
        <li><b>col_to_fill_then_drop:</b> If a stock has a min lower than 0 we must interpolate nulls and then drop the stocks that still have a null in that column.  That is because if we were to simply fill with 0 it would misrepresent that stock, and we may end up with large jumps from negative to positive that increases our variance</li>
        <li><b>col_to_fill:</b> If a stock has a min higher than 0 we can simply interpolate nulls and then set the rest of the nulls to 0</li>
    </ul>
</div>

### More cleaning

#### Fill null ExDividendDate with 1970-01-01

In [21]:
fill_date = pd.to_datetime('1970-01-01')
df_performance = df_performance.reset_index().set_index('date', drop=True)
df_performance['ExDividendDate'] = pd.to_datetime(df_performance['ExDividendDate'].fillna(fill_date))
df_performance['ExDividendDate'].isna().sum()

0

#### Encode ExDividendDate

In [22]:
df_performance['ExDividendDate'] = \
    df_performance['ExDividendDate'].map(lambda x: (x - fill_date).days)

### Execute col_to_fill
> We will use the way we interpolated prices as a reference, and do something similar.  Although we already reindexed, and we don't have to drop symbols here so we can skip those steps.  This will also be converting the data to numeric.  We will also be removing the limit, as some this isn't as important as prices

In [23]:
# Make col_to_fill
col_to_fill = ['sym']
col_to_fill += list(performance_mins[performance_mins >= 0].index)

# Split df_performance into a temp dataframe
df_performance_temp = df_performance[col_to_fill]

# Get all symbol list
all_symbols = df_performance_temp['sym'].unique()

# Make an empty list of all dataframes that will be combined
sym_dfs = []

# For time display
sym_num = len(all_symbols)
times = np.array([])

def cust_interpolate_fill_na(sym):
    """Function to convert columns to numeric and interpolate null values"""
    df = df_performance_temp.query(f"sym == '{sym}'")
    for col in df.columns:
        if col != 'sym':
            df[col] = pd.to_numeric(df[col], errors='coerce').interpolate()
    sym_dfs.append(df)

# Loop through each symbol and apply the above function
for num, sym in enumerate(all_symbols):
    start = time.time()
    cust_interpolate_fill_na(sym)  # Timed function
    end = time.time()
    time_taken = end-start
    times = np.append(times, time_taken)
    print(f"{num+1}/{sym_num} in {time_taken: .2f}s for {sym} :: averaging {times.mean():.2f}s/sym...", end='\r')

# Combine all symbol dataframes, and display info
df_performance_temp =  pd.concat(sym_dfs)

nulls_filled = df_performance[col_to_fill].isna().sum().sum() - \
    df_performance_temp.isna().sum().sum()
print("Null values filled =", nulls_filled)
print("Null values left after filling with 0")

# Fill rest of na with 0
df_performance[col_to_fill] = df_performance_temp.fillna(0)
get_perc_not_null(df_performance)

Null values filled = 1438213 :: averaging 0.03s/sym......
Null values left after filling with 0


FreeCashFlow               75.711
TotalReturn3Yr             77.774
RevenueTTM                 78.602
EnterpriseValue            79.158
TotalDebtToEquity          79.572
ReturnonEquity             83.120
PricePerformanceYTD        84.625
TotalReturn1Yr             88.693
PricePerformance52Weeks    88.693
sym                       100.000
PE                        100.000
PriceToCashFlow           100.000
PriceToSales              100.000
PriceToBook               100.000
DividendYield             100.000
Dividend                  100.000
ExDividendDate            100.000
TotalCash                 100.000
MCAP                      100.000
SharesOutstanding         100.000
VolumeAvg90Day            100.000
dtype: float64

In [24]:
df_performance['sym'].value_counts().min(), df_performance['sym'].value_counts().max()

(339, 339)

### Execute col_to_fill_then_drop
> Interpolate null values then drop stocks that still have nulls in these columns

#### Interpolating the null values

In [25]:
all_symbols = df_performance['sym'].unique()

sym_dfs = []
sym_num = len(all_symbols)
times = np.array([])

def cust_interpolate_fill_na(sym):
    df = df_performance.query(f"sym == '{sym}'")
    for col in df.columns:
        if col != 'sym':
            df[col] = pd.to_numeric(df[col], errors='coerce').interpolate()
    sym_dfs.append(df)

for num, sym in enumerate(all_symbols):
    start = time.time()
    cust_interpolate_fill_na(sym)  # Timed function
    end = time.time()
    time_taken = end-start
    times = np.append(times, time_taken)
    print(f"{num+1}/{sym_num} in {time_taken: .2f}s for {sym} :: averaging {times.mean():.2f}s/sym...", end='\r')

df_performance =  pd.concat(sym_dfs)
df_performance.head()

sym      MCAP  RevenueTTM  TotalCash  ExDividendDate  Dividend  \
date                                                                        
2019-08-09   A 22027.705    5019.000   2155.000               0     0.164   
2019-08-12   A 21758.262    5019.000   2155.000               0     0.164   
2019-08-13   A 21121.110    5019.000   2155.000               0     0.164   
2019-08-14   A 21441.271    5019.000   2155.000               0     0.164   
2019-08-15   A 20332.900    5090.000   1765.000               0     0.164   

            DividendYield  PriceToBook  PriceToSales  EnterpriseValue  ...  \
date                                                                   ...   
2019-08-09          0.944        4.298         4.402  21670704590.000  ...   
2019-08-12          0.956        4.246         4.348  21401262240.000  ...   
2019-08-13          0.985        4.121         4.221  20764110330.000  ...   
2019-08-14          0.970        4.184         4.285  21084271240.000  ...   
2019-08-15          1.000        4.283         4.068  20365900000.000  ...   

               PE  ReturnonEquity  FreeCashFlow  TotalDebtToEquity  \
date                                                                 
2019-08-09 19.800          23.160       887.000             37.290   
2019-08-12 19.200          23.160       887.000             37.290   
2019-08-13 19.500          23.160       887.000             37.290   
2019-08-14 18.900          23.160       887.000             37.290   
2019-08-15 20.800          22.018       918.000             36.935   

            PricePerformance52Weeks  PricePerformanceYTD  TotalReturn1Yr  \
date                                                                       
2019-08-09                    3.131                3.009           4.044   
2019-08-12                    3.592                1.749           4.508   
2019-08-13                    0.558               -1.230           1.448   
2019-08-14                    2.578                0.267           3.485   
2019-08-15                   -1.738               -2.772          -0.869   

            TotalReturn3Yr  SharesOutstanding  VolumeAvg90Day  
date                                                           
2019-08-09          14.794            315.993           2.593  
2019-08-12          14.099            315.993           2.585  
2019-08-13          12.479            315.993           2.605  
2019-08-14          13.044            315.993           2.616  
2019-08-15          11.891            315.993           2.631  

[5 rows x 21 columns]

#### Dropping the stocks that still have nulls

In [26]:
# Make for reference, to see how many stocks get dropped
num_stocks = df_performance.reset_index()['sym'].nunique()

# Create col_to_fill_then_drop
col_to_fill_then_drop = ['sym']
col_to_fill_then_drop += list(performance_mins[performance_mins < 0].index)

# Make for reference, to see how many stocks get dropped
num_stocks = df_performance.reset_index()['sym'].nunique()

# Create col_to_fill_then_drop
col_to_fill_then_drop = ['sym']
col_to_fill_then_drop += list(performance_mins[performance_mins < 0].index)

# Create a isna column with counts of NaNs in the given columns
df_performance['isna'] = df_performance[col_to_fill_then_drop].isna().sum(axis=1)

# Create a variable drop_symbol with a list of symbols that need to be dropped
var = df_performance.groupby('sym').sum()
drop_symbol = var[var['isna'] > 0].index

# Drop the rows with the given symbols
df_performance = df_performance.reset_index() \
    [~df_performance.reset_index()['sym'].isin(drop_symbol)]

# Drop isna column
df_performance.drop('isna', axis=1, inplace=True)

# Displays
perc_removed = 100 - (df_performance['sym'].nunique() / num_stocks)*100
print(f"Removed {perc_removed: .2f}% of the stocks from the data")
print(df_performance['sym'].nunique(), "Stocks left vs", num_stocks, "before")
get_perc_not_null(df_performance)

Removed  51.53% of the stocks from the data
2958 Stocks left vs 6103 before


date                      100.000
TotalReturn3Yr            100.000
TotalReturn1Yr            100.000
PricePerformanceYTD       100.000
PricePerformance52Weeks   100.000
TotalDebtToEquity         100.000
FreeCashFlow              100.000
ReturnonEquity            100.000
PE                        100.000
PriceToCashFlow           100.000
EnterpriseValue           100.000
PriceToSales              100.000
PriceToBook               100.000
DividendYield             100.000
Dividend                  100.000
ExDividendDate            100.000
TotalCash                 100.000
RevenueTTM                100.000
MCAP                      100.000
sym                       100.000
SharesOutstanding         100.000
VolumeAvg90Day            100.000
dtype: float64

### Add price to performance

In [27]:
df_performance = df_performance.merge(df_prices.reset_index(), on=['date', 'sym'])
df_performance.isna().sum().sum()

0

### Apply Splits

In [28]:
fresh_df_performance = df_performance.copy(deep=True)
#df_performance = fresh_df_performance.copy(deep=True)

In [29]:
# Make est_price_unsplit column for calculating how far off the price is than
# make an price_est_vs_actual for a reference to calculate if a split is
# expected or not
df_performance['est_price_unsplit'] = df_performance['MCAP'] / df_performance['SharesOutstanding']
df_performance['price_est_vs_actual'] = \
    df_performance['est_price_unsplit'] / df_performance['price']
df_performance['split_expected'] = \
    df_performance['price_est_vs_actual'] \
        .map(lambda x:True if ((x > 1.5) | (x < 0.50)) else False)

# df_performance.reset_index(inplace=True) ???

In [30]:
# df_test = df_performance.loc[((df_performance['MCAP'] > 0) & (df_performance['SharesOutstanding'] > 0) & (df_performance['price'] > 0))].sort_values(by='price_est_vs_actual')
# df_test = df_test.loc[df_test['date'] == '2019-08-09']
# df_test.groupby('sym')['price_est_vs_actual', 'price', 'est_price_unsplit', 'MCAP', 'SharesOutstanding'].mean().sort_values(by='price_est_vs_actual').head(50).to_csv('./anomaly.csv')

In [31]:
for idx, row in df_splits.iterrows():
    sym = row['sym']
    date = row['execute']
    apply = row['apply']
    row_ids = ((df_performance['sym'] == sym) & (df_performance['date'] < pd.to_datetime(date)))
    df_performance.loc[row_ids, 'price'] = \
        df_performance.loc[row_ids, 'price'] / apply
    
    date += pd.to_timedelta('1d')
    row_ids = ((df_performance['sym'] == sym) & (df_performance['date'] < pd.to_datetime(date)))
    row_ids = ((df_performance['sym'] == sym) & (df_performance['date'] < pd.to_datetime(date)))
    df_performance.loc[row_ids, 'split_expected'] = False

In [32]:
view_cols = ['date', 'sym', 'MCAP', 'SharesOutstanding', 'price',
             'est_price_unsplit', 'price_est_vs_actual', 'split_expected']
df_performance.query("sym == 'AAPL'").query("date <= '2020-09-01'") \
    .tail(3)[view_cols].set_index(['date', 'sym'], drop=True)

MCAP  SharesOutstanding   price  est_price_unsplit  \
date       sym                                                              
2020-08-28 AAPL 2142140.858           4275.634 124.725            501.011   
2020-08-31 AAPL  534667.717          17102.536 129.040             31.262   
2020-09-01 AAPL 2211197.954          17102.536 134.180            129.291   

                 price_est_vs_actual  split_expected  
date       sym                                        
2020-08-28 AAPL                1.004           False  
2020-08-31 AAPL                0.242           False  
2020-09-01 AAPL                0.964           False

<div class="alert alert-info shadow">
  <strong>After the split application:</strong> <br></br>
    08-31 was a split day from our split data,  split_expected was true before because of the differences of price and our estimated price.  This is because the <b>MCAP</b> is not accurate on a split day, so we cannot trust that or <b>SharesOutstanding</b> to make accurate predictions, or for calculating actual splits as they are not accurate enough to devise a 1 to 1.05 split.<ul>
        <li>We will be removing stocks that still have splits expected due to the lack of ability to calculate the split</li>
    <li>We will also be dropping <b>MCAP</b> <b>SharesOutstanding</b> and the other columns we created for split calculations</li>
    </ul>
</div>

In [33]:
drop_symbols = df_performance.query("split_expected")['sym'].unique()

# Drop the rows with the given symbols
df_performance = df_performance[~df_performance['sym'].isin(drop_symbols)]

# Drop the created columns, mcap, and shares outstanding
drop_cols = ['est_price_unsplit', 'price_est_vs_actual', 'split_expected', 'SharesOutstanding', 'MCAP']
df_performance.drop(drop_cols, axis=1, inplace=True)

### Separate out penny stocks

In [34]:
df_penny_performance = df_performance.loc[df_performance['price'] < 1]
df_performance = df_performance.drop(df_penny_performance.index)

### View loss

In [35]:
print(df_performance.reset_index()['sym'].nunique(), "/", num_stocks)
perc_removed = 100 - (df_performance['sym'].nunique() / num_stocks)*100
print(f"Removed {perc_removed: .2f}% of the stocks from the data")

2225 / 6103
Removed  63.54% of the stocks from the data


<div class="alert alert-info shadow">
  <strong>63.5%</strong> Of our stocks have now been removed in the hopes of more accurate predictions.  We are left with 2,229 stocks.
</div>

## Pickle

In [36]:
df_performance.to_pickle('./data/performance.pkl')
df_penny_performance.to_pickle('./data/penny_performance.pkl')

In [37]:
(pd.read_pickle('./data/performance.pkl') != df_performance).sum().sum()

0

In [38]:
(pd.read_pickle('./data/penny_performance.pkl') != df_penny_performance).sum().sum()

0

## Post to sql

In [39]:
db_clean = DataBase('stock_cleaned', 14333, 'localhost')

In [40]:
db_clean.pull_names_as_dataframes(['penny_performance', 'performance'])
db_clean.frames['performance']

sym        date      price RevenueTTM TotalCash ExDividendDate  \
0          A  2019-08-09  68.640000       5019      2155              0   
1          A  2019-08-12  66.780000       5019      2155              0   
2          A  2019-08-13  67.640000       5019      2155              0   
3          A  2019-08-14  65.590000       5019      2155              0   
4          A  2019-08-15  69.590000       5090      1765              0   
...      ...         ...        ...        ...       ...            ...   
727147  ZYXI  2020-11-23  13.620000     68.679    41.229              0   
727148  ZYXI  2020-11-24  13.440000     68.679    41.229              0   
727149  ZYXI  2020-11-25  13.390000     68.679    41.229              0   
727150  ZYXI  2020-11-26  13.390000     68.679    41.229              0   
727151  ZYXI  2020-11-27  13.870000     68.679    41.229              0   

       Dividend DividendYield PriceToBook PriceToSales  ... PriceToCashFlow  \
0         0.164       0.94402     4.29809      4.40182  ...        16.57368   
1         0.164       0.95571     4.24551      4.34797  ...        16.37095   
2         0.164       0.98454     4.12119      4.22065  ...        15.89156   
3         0.164       0.96984     4.18366      4.28463  ...        16.13245   
4         0.164       1.00015     4.28332      4.06782  ...        16.05056   
...         ...           ...         ...          ...  ...             ...   
727147        0             0     8.58335      6.54801  ...        39.23495   
727148        0             0     8.58335      6.54801  ...        39.23495   
727149        0             0     8.59615      6.55777  ...        39.29342   
727150        0             0     8.58335      6.54801  ...        39.23495   
727151        0             0     8.58335      6.54801  ...        39.23495   

          PE ReturnonEquity FreeCashFlow TotalDebtToEquity  \
0       19.8        23.1598          887          37.29007   
1       19.2        23.1598          887          37.29007   
2       19.5        23.1598          887          37.29007   
3       18.9        23.1598          887          37.29007   
4       20.8        22.0179          918          36.93453   
...      ...            ...          ...               ...   
727147  45.4        32.8813        3.825          17.03027   
727148  44.8        32.8813        3.825          17.03027   
727149  44.6        32.8813        3.825          17.03027   
727150  44.6        32.8813        3.825          17.03027   
727151  46.2        32.8813        3.825          17.03027   

       PricePerformance52Weeks PricePerformanceYTD TotalReturn1Yr  \
0                      3.13149             3.00919        4.04369   
1                      3.59191             1.74918        4.50818   
2                      0.55841            -1.23036        1.44784   
3                       2.5781             0.26682         3.4854   
4                     -1.73783            -2.77201        -0.8687   
...                        ...                 ...            ...   
727147                48.28729            70.52097       48.28729   
727148                48.28729            70.52097       48.28729   
727149                50.67265             70.7751       50.67265   
727150                25.18657            70.52097       25.18657   
727151                25.18657            70.52097       25.18657   

       TotalReturn3Yr VolumeAvg90Day  
0            14.79356        2.59274  
1            14.09856        2.58484  
2             12.4788        2.60485  
3            13.04428        2.61616  
4            11.89052        2.63078  
...               ...            ...  
727147       66.20773        0.91285  
727148       66.20773        0.91285  
727149       66.29025        0.84596  
727150       66.20773        0.83464  
727151       66.20773        0.83464  

[727152 rows x 21 columns]

#### Performance

In [41]:
# Make query list
query_list = []
for idx, row in df_performance.iterrows():
    query_list.append(tuple(row.to_dict().values()))

# Execute
db_clean.cur.executemany("""
REPLACE INTO performance (date, sym, RevenueTTM, TotalCash, ExDividendDate,
                    Dividend, DividendYield, PriceToBook, PriceToSales,
                    EnterpriseValue, PriceToCashFlow, PE, ReturnonEquity,
                    FreeCashFlow, TotalDebtToEquity, PricePerformance52Weeks,
                    PricePerformanceYTD, TotalReturn1Yr, TotalReturn3Yr,
                    VolumeAvg90Day, price)
VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)""",
      query_list)

# commit
db_clean.conn.commit()

#### Penny

In [42]:
# Make query list
query_list_penny = []
for idx, row in df_penny_performance.iterrows():
    query_list_penny.append(tuple(row.to_dict().values()))
    
db_clean.cur.executemany("""
REPLACE INTO penny_performance (date, sym, RevenueTTM, TotalCash, ExDividendDate,
                    Dividend, DividendYield, PriceToBook, PriceToSales,
                    EnterpriseValue, PriceToCashFlow, PE, ReturnonEquity,
                    FreeCashFlow, TotalDebtToEquity, PricePerformance52Weeks,
                    PricePerformanceYTD, TotalReturn1Yr, TotalReturn3Yr,
                    VolumeAvg90Day, price)
VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)""",
      query_list_penny)

db_clean.conn.commit()

> Mostly equal other than rounding errors.  Good to go!

# Company

## Clean for new sql

In [43]:
df_company = db.frames['company'].copy(deep=True)

# Resetting date index and dropping index
df_company.set_index('sym', inplace=True, drop=True)

df_company.drop('date', inplace=True, axis=1)

df_company.replace(['', '--'], np.NaN, inplace=True)
df_company.head()

company securityType                  sector  \
sym                                                                   
A     Agilent Technologies Inc           CS             Health Care   
AA                  Alcoa Corp           CS               Materials   
AABA                Altaba Inc           CS  Consumer Discretionary   
AAC           AAC Holdings Inc           CS             Health Care   
AACG     ATA Creativity Global           DR  Consumer Discretionary   

                                industry                         subIndustry  \
sym                                                                            
A         Life Sciences Tools & Services      Life Sciences Tools & Services   
AA                       Metals & Mining                            Aluminum   
AABA  Internet & Direct Marketing Retail  Internet & Direct Marketing Retail   
AAC     Health Care Providers & Services              Health Care Facilities   
AACG       Diversified Consumer Services                  Education Services   

     Exchange DJIA NASDAQ100 SP500  
sym                                 
A        NYSE    0         0     1  
AA       NYSE    0         0     0  
AABA   NASDAQ    0         0     0  
AAC      NYSE    0         0     0  
AACG   NASDAQ    0         0     0

In [44]:
df_company.nunique()

company         7366
securityType       6
sector            11
industry          69
subIndustry      161
Exchange           4
DJIA               2
NASDAQ100          2
SP500              2
dtype: int64

In [45]:
len(df_company)

7936

In [46]:
df_company.isna().sum()

company           0
securityType      6
sector          923
industry        923
subIndustry     923
Exchange         37
DJIA              0
NASDAQ100         0
SP500             0
dtype: int64

In [47]:
df_company = df_company.fillna('unknown')

In [48]:
df_company.isna().sum()

company         0
securityType    0
sector          0
industry        0
subIndustry     0
Exchange        0
DJIA            0
NASDAQ100       0
SP500           0
dtype: int64

In [49]:
all_symbols = df_performance.reset_index()['sym'].unique()

In [50]:
df_company = df_company[df_company.index.isin(all_symbols)]

In [51]:
df_company.to_pickle('./data/company.pkl')

# Analyst

In [52]:
# Cleaning first two blank rows
df_analyst = db.frames['analyst'].copy(deep=True).drop([0, 1])

# Resetting date index and dropping index
df_analyst.set_index('date', inplace=True, drop=True)

# Dropping null date values
df_analyst = df_analyst.loc[df_analyst.index.notnull()]

df_analyst.replace(['', '--'], np.NaN, inplace=True)

df_analyst.head()

sym Argus StarMine ColumbineCapital FordEquityResearch ISSEVA  \
date                                                                       
2019-06-04   A   NaN      NaN                1                  3     70   
2019-06-08   A   NaN      NaN                1                  3     70   
2019-06-17   A   NaN      9.4                1                  3     70   
2019-06-19   A   NaN      9.4                1                  3     70   
2019-06-20   A   NaN        7                1                  3     70   

           Jefferson McLeanCapital NedDavis TradingCentral Zacks SPValuation  \
date                                                                           
2019-06-04       Buy       Neutral  NEUTRAL            NaN  HOLD          92   
2019-06-08       Buy       Neutral  NEUTRAL            NaN   BUY          93   
2019-06-17         1             3  NEUTRAL              3   BUY          93   
2019-06-19         1             3  NEUTRAL              3  HOLD          93   
2019-06-20         1             3  NEUTRAL              3  HOLD          93   

           SPQuality SPGrowthStability SPFinancialHealth ThomsonReutersVerus  \
date                                                                           
2019-06-04        75                 7                85                HOLD   
2019-06-08        74                 6                85                HOLD   
2019-06-17        74                 7                85                HOLD   
2019-06-19        75                 7                85                HOLD   
2019-06-20        75                 7                85                HOLD   

           ThomsonReutersIBES  
date                           
2019-06-04              1.714  
2019-06-08                1.8  
2019-06-17              1.813  
2019-06-19              1.813  
2019-06-20              1.813

In [53]:
perf_symbols = df_prices['sym'].unique()
df_analyst = df_analyst.loc[df_analyst['sym'].isin(perf_symbols)]
test_analyst = df_analyst.reset_index().set_index(['date', 'sym'])
test_performance = df_performance.reset_index().set_index(['date', 'sym'])
df_analyst = pd.DataFrame(test_analyst, index=test_performance.index)
df_analyst.shape, df_performance.shape

((730822, 16), (730822, 21))

## Checking for null values

In [54]:
get_perc_not_null(df_analyst)

Argus                 16.139
NedDavis              24.177
TradingCentral        39.730
ColumbineCapital      56.147
McLeanCapital         59.815
Jefferson             64.994
ISSEVA                74.702
Zacks                 77.436
StarMine              79.656
FordEquityResearch    80.454
ThomsonReutersIBES    82.482
ThomsonReutersVerus   89.165
SPFinancialHealth     90.321
SPGrowthStability     91.766
SPQuality             91.780
SPValuation           91.783
dtype: float64

> Here you can see the percentage of a column that is not null, Argus is in less than one percent of the data.  We will need to convene a method to deal with the null values.

>  The method for filling null values will be a front fill, and a 0 based back fill grouped by stock.  For example if a stock has not been rated by a company at all it will be continuously 0, and if that analyst suddenly rating a stock after no having rated it is significant an algorithm should pick up on that.

In [55]:
test_fill_na_dict = {
    'sym': ['a', 'a', 'a', 'a', 'a'],
    'random_analyst': [np.NaN, np.NaN, 3, np.NaN, 5],
    'randon_analyst2': [3, 4, 8, 6, 5],
}

test_df2 = pd.DataFrame(test_fill_na_dict)
display(test_df2.head())

test_df2.fillna(method='ffill').fillna(0)

sym  random_analyst  randon_analyst2
0   a             nan                3
1   a             nan                4
2   a           3.000                8
3   a             nan                6
4   a           5.000                5

sym  random_analyst  randon_analyst2
0   a           0.000                3
1   a           0.000                4
2   a           3.000                8
3   a           3.000                6
4   a           5.000                5

<div class="alert alert-info shadow">
  The <strong>fillna(method="ffill").fillna(0)</strong> does exactly that by first filling nulls with the front value, and then the second fillna fills the rest of the nulls with 0.  We will now use it on the entire dataframe.
</div>

## Filling null values

In [56]:
df_analyst = df_analyst.fillna(method='ffill').fillna(0)

In [57]:
# Checking that all null values are filled
df_analyst.isna().sum()

Argus                  0
StarMine               0
ColumbineCapital       0
FordEquityResearch     0
ISSEVA                 0
Jefferson              0
McLeanCapital          0
NedDavis               0
TradingCentral         0
Zacks                  0
SPValuation            0
SPQuality              0
SPGrowthStability      0
SPFinancialHealth      0
ThomsonReutersVerus    0
ThomsonReutersIBES     0
dtype: int64

In [58]:
len(df_analyst)

730822

<div class="alert alert-success shadow">
  <strong>Success!</strong> No more null values in analyst, and it has 624,265 rows.  Now we need to go about making the different columns numerical.  There are two different methods we could use for this:
    <ul>
        <li>Manually take the unique values for each column and <b>mapping</b> an arbitrary number to represent things like 'BUY', 'HOLD', 'SELL BIGLY'</li>
        <li>Use <b>NLP</b> to vectorize each of the columns</li><ul> <br> We will be using the first method, as we already have the mapping ready from a previous analysis.

</div>

## Mapping

In [59]:
analyst_mapping = {
    'Argus': {0:0, 'Buy':5, 'Hold':3, 'Sell':1},
    'TradingCentral': {'Buy':5, '1':1, '3':3, '5':5, 0:0, 'Sell':1},
    'Jefferson': {'Buy':5, '1':1, '3':3, '5':5, 'Hold':3, 'Sell':1},
    'McLeanCapital': {'Buy':5, '1':1, '3':3, '5':5, 'Neutral':3, 'Sell':1},
    'NedDavis': {'NEUTRAL':3, 'BUY':5, 'SELL':1},
    'Zacks': {'STRONG BUY':6, 'BUY':5, 'OUTPERFORM':5,
             'NEUTRAL':3, 'HOLD':3, 'SELL':2, 'UNDERPERFORM':2,
             'STRONG SELL':1, 'NONE':0},
    'ThomsonReutersVerus': {'HOLD':3, 'SELL':1, 'BUY':5}
}
for col, mapping in analyst_mapping.items():
    df_analyst[col] = df_analyst[col].map(mapping)
    
df_analyst.isna().sum()

Argus                  0
StarMine               0
ColumbineCapital       0
FordEquityResearch     0
ISSEVA                 0
Jefferson              0
McLeanCapital          0
NedDavis               0
TradingCentral         0
Zacks                  0
SPValuation            0
SPQuality              0
SPGrowthStability      0
SPFinancialHealth      0
ThomsonReutersVerus    0
ThomsonReutersIBES     0
dtype: int64

## Convert strings to float

In [60]:
for col in df_analyst.columns:
    if col != 'sym':
        print(col)
        df_analyst[col] = pd.to_numeric(df_analyst[col])

Argus
StarMine
ColumbineCapital
FordEquityResearch
ISSEVA
Jefferson
McLeanCapital
NedDavis
TradingCentral
Zacks
SPValuation
SPQuality
SPGrowthStability
SPFinancialHealth
ThomsonReutersVerus
ThomsonReutersIBES


In [61]:
df_analyst['TradingCentral'].unique()

array([3, 1, 5], dtype=int64)

In [62]:
df_analyst.describe()

Argus   StarMine  ColumbineCapital  FordEquityResearch     ISSEVA  \
count 730822.000 730822.000        730822.000          730822.000 730822.000   
mean       4.287      4.839             1.905               2.827     51.064   
std        0.995      2.933             0.783               0.830     29.118   
min        0.000     -8.900             1.000               1.000      0.000   
25%        3.000      2.200             1.000               2.000     26.000   
50%        5.000      4.700             2.000               3.000     52.000   
75%        5.000      7.400             3.000               3.000     76.000   
max        5.000     10.000             3.000               5.000    100.000   

       Jefferson  McLeanCapital   NedDavis  TradingCentral      Zacks  \
count 730822.000     730822.000 730822.000      730822.000 730822.000   
mean       3.430          2.871      3.092           2.917      3.167   
std        1.494          0.620      0.883           1.270      0.914   
min        1.000          1.000      1.000           1.000      1.000   
25%        3.000          3.000      3.000           3.000      3.000   
50%        3.000          3.000      3.000           3.000      3.000   
75%        5.000          3.000      3.000           3.000      3.000   
max        5.000          5.000      5.000           5.000      6.000   

       SPValuation  SPQuality  SPGrowthStability  SPFinancialHealth  \
count   730822.000 730822.000         730822.000         730822.000   
mean        53.423     54.687             48.323             52.800   
std         27.643     27.642             28.303             26.490   
min          0.000      0.000              0.000              0.000   
25%         31.000     32.000             24.000             31.000   
50%         55.000     57.000             49.000             53.000   
75%         77.000     78.000             72.000             75.000   
max         99.000     99.000             99.000             99.000   

       ThomsonReutersVerus  ThomsonReutersIBES  
count           730822.000          730822.000  
mean                 2.997               2.285  
std                  1.134               0.550  
min                  1.000               1.000  
25%                  3.000               2.000  
50%                  3.000               2.235  
75%                  3.000               2.667  
max                  5.000               5.000

In [63]:
df_analyst.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 730822 entries, (Timestamp('2019-08-09 00:00:00'), 'A') to (Timestamp('2020-12-01 00:00:00'), 'ZYXI')
Data columns (total 16 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Argus                730822 non-null  int64  
 1   StarMine             730822 non-null  float64
 2   ColumbineCapital     730822 non-null  int64  
 3   FordEquityResearch   730822 non-null  int64  
 4   ISSEVA               730822 non-null  int64  
 5   Jefferson            730822 non-null  int64  
 6   McLeanCapital        730822 non-null  int64  
 7   NedDavis             730822 non-null  int64  
 8   TradingCentral       730822 non-null  int64  
 9   Zacks                730822 non-null  int64  
 10  SPValuation          730822 non-null  int64  
 11  SPQuality            730822 non-null  int64  
 12  SPGrowthStability    730822 non-null  int64  
 13  SPFinancialHealth    730822 non-null  int64  

## Add to sql server

In [64]:
db_clean = DataBase('stock_cleaned', 14333, 'localhost')
db_clean.pull_names_as_dataframes(['analyst'])
db_clean.frames['analyst']

sym        date  Argus  StarMine  ColumbineCapital  \
0          A  2019-08-09      0     5.800                 1   
1          A  2019-08-12      0     8.300                 1   
2          A  2019-08-13      0     8.300                 1   
3          A  2019-08-14      0     4.200                 1   
4          A  2019-08-15      0     4.200                 1   
...      ...         ...    ...       ...               ...   
727147  ZYXI  2020-11-23      5     0.600                 2   
727148  ZYXI  2020-11-24      5     0.600                 2   
727149  ZYXI  2020-11-25      5     0.600                 2   
727150  ZYXI  2020-11-26      5     0.600                 2   
727151  ZYXI  2020-11-27      5     0.600                 2   

        FordEquityResearch  ISSEVA  Jefferson  McLeanCapital  NedDavis  \
0                        3      78          1              3         3   
1                        3      78          1              3         3   
2                        3      73          1              3         3   
3                        3      73          1              3         3   
4                        3      73          1              3         3   
...                    ...     ...        ...            ...       ...   
727147                   3      44          5              3         3   
727148                   3      44          5              3         3   
727149                   3      43          5              3         3   
727150                   3      42          5              3         3   
727151                   3      42          5              3         3   

        TradingCentral  Zacks  SPValuation  SPQuality  SPGrowthStability  \
0                    3      3           93         73                  6   
1                    3      5           93         73                  6   
2                    3      3           93         73                  7   
3                    3      3           93         73                  7   
4                    3      3           93         74                  7   
...                ...    ...          ...        ...                ...   
727147               1      2           61         68                 15   
727148               1      2           61         68                 15   
727149               1      2           62         68                 15   
727150               1      2           61         69                 15   
727151               1      2           61         69                 15   

        SPFinancialHealth  ThomsonReutersVerus  ThomsonReutersIBES  
0                      84                    3               2.000  
1                      84                    3               2.000  
2                      85                    3               2.000  
3                      85                    3               2.000  
4                      85                    3               2.000  
...                   ...                  ...                 ...  
727147                 70                    3               1.800  
727148                 70                    3               1.800  
727149                 70                    3               1.800  
727150                 70                    3               1.800  
727151                 70                    3               1.800  

[727152 rows x 18 columns]

In [65]:
query_list = []
for idx, row in df_analyst.reset_index().iterrows():
    query_list.append(tuple(row.to_dict().values()))

db_clean = DataBase('stock_cleaned', 14333, 'localhost')
db_clean.cur.executemany("""
REPLACE INTO analyst (date, sym, Argus, StarMine, ColumbineCapital,
                         FordEquityResearch, ISSEVA, Jefferson, McLeanCapital,
                         NedDavis, TradingCentral, Zacks, SPValuation,
                         SPQuality, SPGrowthStability, SPFinancialHealth,
                         ThomsonReutersVerus, ThomsonReutersIBES)
VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)""",
      query_list)
db_clean.conn.commit()
db_clean.conn.close()

## Pickle

In [66]:
df_analyst.to_pickle('./data/analyst.pkl')

# Combine company, analyst and performance

In [67]:
df_company.drop('subIndustry', axis=1, inplace=True)

In [68]:
def combine_df(df, company, analyst):
    ohe_columns = list(company.columns)
    ohe_columns.remove('company')
    df = df.reset_index().set_index(['date', 'sym'], drop=True)
    analyst = analyst.reset_index().set_index(['date', 'sym'], drop=True)
    df_new = MLFrame(df.join(df_company).join(analyst))
    df_new.drop('company', inplace=True, axis=1)
    df_new = df_new.one_hot_encode(columns=ohe_columns)
    return df_new    

## Performance

In [69]:
df_performance_combined = combine_df(df_performance, df_company, df_analyst)
df_performance_combined.head()

Added categorical columns
3 -> Dividend
4 -> Exchange
5 -> securityType
10 -> sector
68 -> industry


index  RevenueTTM  TotalCash  ExDividendDate  Dividend  \
date       sym                                                           
2019-08-09 A        0    5019.000   2155.000               0     0.164   
2019-08-12 A        1    5019.000   2155.000               0     0.164   
2019-08-13 A        2    5019.000   2155.000               0     0.164   
2019-08-14 A        3    5019.000   2155.000               0     0.164   
2019-08-15 A        4    5090.000   1765.000               0     0.164   

                DividendYield  PriceToBook  PriceToSales  EnterpriseValue  \
date       sym                                                              
2019-08-09 A            0.944        4.298         4.402  21670704590.000   
2019-08-12 A            0.956        4.246         4.348  21401262240.000   
2019-08-13 A            0.985        4.121         4.221  20764110330.000   
2019-08-14 A            0.970        4.184         4.285  21084271240.000   
2019-08-15 A            1.000        4.283         4.068  20365900000.000   

                PriceToCashFlow  ...  industry_Transportation Infrastructure  \
date       sym                   ...                                           
2019-08-09 A             16.574  ...                                       0   
2019-08-12 A             16.371  ...                                       0   
2019-08-13 A             15.892  ...                                       0   
2019-08-14 A             16.132  ...                                       0   
2019-08-15 A             16.051  ...                                       0   

                industry_Water Utilities  \
date       sym                             
2019-08-09 A                           0   
2019-08-12 A                           0   
2019-08-13 A                           0   
2019-08-14 A                           0   
2019-08-15 A                           0   

                industry_Wireless Telecommunication Services  Exchange_NASDAQ  \
date       sym                                                                  
2019-08-09 A                                               0                0   
2019-08-12 A                                               0                0   
2019-08-13 A                                               0                0   
2019-08-14 A                                               0                0   
2019-08-15 A                                               0                0   

                Exchange_NYSE  Exchange_OTC  Exchange_unknown  DJIA_1  \
date       sym                                                          
2019-08-09 A                1             0                 0       0   
2019-08-12 A                1             0                 0       0   
2019-08-13 A                1             0                 0       0   
2019-08-14 A                1             0                 0       0   
2019-08-15 A                1             0                 0       0   

                NASDAQ100_1  SP500_1  
date       sym                        
2019-08-09 A              0        1  
2019-08-12 A              0        1  
2019-08-13 A              0        1  
2019-08-14 A              0        1  
2019-08-15 A              0        1  

[5 rows x 126 columns]

## Penny

In [70]:
df_penny_performance_combined = combine_df(df_penny_performance, df_company, df_analyst)
df_penny_performance_combined.head()

Added categorical columns
3 -> Dividend
3 -> securityType
3 -> Exchange
10 -> sector
46 -> industry


index  RevenueTTM  TotalCash  ExDividendDate  Dividend  \
date       sym                                                           
2019-08-09 AAC    339     275.946     17.869               0     0.000   
2019-08-12 AAC    340     275.946     17.869               0     0.000   
2019-08-13 AAC    341     275.946     17.869               0     0.000   
2019-08-14 AAC    342     275.946     17.869               0     0.000   
2019-08-15 AAC    343     275.946     17.869               0     0.000   

                DividendYield  PriceToBook  PriceToSales  EnterpriseValue  \
date       sym                                                              
2019-08-09 AAC          0.000        0.502         0.078    405338418.000   
2019-08-12 AAC          0.000        0.506         0.079    405516013.200   
2019-08-13 AAC          0.000        0.471         0.073    403996587.600   
2019-08-14 AAC          0.000        0.466         0.072    403786926.600   
2019-08-15 AAC          0.000        0.385         0.060    400331220.000   

                PriceToCashFlow  ...  \
date       sym                   ...   
2019-08-09 AAC            0.000  ...   
2019-08-12 AAC            0.000  ...   
2019-08-13 AAC            0.000  ...   
2019-08-14 AAC            0.000  ...   
2019-08-15 AAC            0.000  ...   

                industry_Real Estate Management & Development  \
date       sym                                                  
2019-08-09 AAC                                              0   
2019-08-12 AAC                                              0   
2019-08-13 AAC                                              0   
2019-08-14 AAC                                              0   
2019-08-15 AAC                                              0   

                industry_Road & Rail  industry_Software  \
date       sym                                            
2019-08-09 AAC                     0                  0   
2019-08-12 AAC                     0                  0   
2019-08-13 AAC                     0                  0   
2019-08-14 AAC                     0                  0   
2019-08-15 AAC                     0                  0   

                industry_Specialty Retail  \
date       sym                              
2019-08-09 AAC                          0   
2019-08-12 AAC                          0   
2019-08-13 AAC                          0   
2019-08-14 AAC                          0   
2019-08-15 AAC                          0   

                industry_Technology Hardware, Storage & Peripherals  \
date       sym                                                        
2019-08-09 AAC                                                  0     
2019-08-12 AAC                                                  0     
2019-08-13 AAC                                                  0     
2019-08-14 AAC                                                  0     
2019-08-15 AAC                                                  0     

                industry_Textiles, Apparel & Luxury Goods  \
date       sym                                              
2019-08-09 AAC                                          0   
2019-08-12 AAC                                          0   
2019-08-13 AAC                                          0   
2019-08-14 AAC                                          0   
2019-08-15 AAC                                          0   

                industry_Trading Companies & Distributors  Exchange_NASDAQ  \
date       sym                                                               
2019-08-09 AAC                                          0                0   
2019-08-12 AAC                                          0                0   
2019-08-13 AAC                                          0                0   
2019-08-14 AAC                                          0                0   
2019-08-15 AAC                                          0                0   

                E

## Write pickles

In [71]:
df_performance_combined.to_pickle('./data/performance_combined.pkl')
df_penny_performance_combined.to_pickle('./data/performance_penny_combined.pkl')

# Post to firebase
> Need to add updating

In [72]:
from db import FireBase
import json
import os

In [73]:
os.listdir('./data')

['analyst.pkl',
 'company.pkl',
 'modeling',
 'penny_performance.pkl',
 'performance.pkl',
 'performance_combined.pkl',
 'performance_penny_combined.pkl',
 'prices.pkl']

In [74]:
df_performance = pd.read_pickle('./data/performance.pkl')
df_penny_performance = pd.read_pickle('./data/penny_performance.pkl')

In [75]:
df_post = pd.concat([df_performance, df_penny_performance])
df_post = df_post.set_index(['date'], drop=True)
df_post.head()

sym  RevenueTTM  TotalCash  ExDividendDate  Dividend  \
date                                                              
2019-08-09   A    5019.000   2155.000               0     0.164   
2019-08-12   A    5019.000   2155.000               0     0.164   
2019-08-13   A    5019.000   2155.000               0     0.164   
2019-08-14   A    5019.000   2155.000               0     0.164   
2019-08-15   A    5090.000   1765.000               0     0.164   

            DividendYield  PriceToBook  PriceToSales  EnterpriseValue  \
date                                                                    
2019-08-09          0.944        4.298         4.402  21670704590.000   
2019-08-12          0.956        4.246         4.348  21401262240.000   
2019-08-13          0.985        4.121         4.221  20764110330.000   
2019-08-14          0.970        4.184         4.285  21084271240.000   
2019-08-15          1.000        4.283         4.068  20365900000.000   

            PriceToCashFlow     PE  ReturnonEquity  FreeCashFlow  \
date                                                               
2019-08-09           16.574 19.800          23.160       887.000   
2019-08-12           16.371 19.200          23.160       887.000   
2019-08-13           15.892 19.500          23.160       887.000   
2019-08-14           16.132 18.900          23.160       887.000   
2019-08-15           16.051 20.800          22.018       918.000   

            TotalDebtToEquity  PricePerformance52Weeks  PricePerformanceYTD  \
date                                                                          
2019-08-09             37.290                    3.131                3.009   
2019-08-12             37.290                    3.592                1.749   
2019-08-13             37.290                    0.558               -1.230   
2019-08-14             37.290                    2.578                0.267   
2019-08-15             36.935                   -1.738               -2.772   

            TotalReturn1Yr  TotalReturn3Yr  VolumeAvg90Day  price  
date                                                               
2019-08-09           4.044          14.794           2.593 68.640  
2019-08-12           4.508          14.099           2.585 66.780  
2019-08-13           1.448          12.479           2.605 67.640  
2019-08-14           3.485          13.044           2.616 65.590  
2019-08-15          -0.869          11.891           2.631 69.590

In [76]:
fb = FireBase()
post = {}
all_symbols = df_post['sym'].unique()
sym_df = df_post.groupby('sym')

x = 0
for sym in all_symbols:
    x += 1
    post[sym] = sym_df.get_group(sym).drop('sym', axis=1).reset_index(drop=True).to_dict()
    if x == 100:
        print('posting ->', sym, end='\r')
        req = fb.refresh_all_data(post)
        print('posted ->', sym, end='\r')
        post = {}
        x = 0

req = fb.refresh_all_data(post)

In [77]:
date_range = df_post.query("sym == 'AAPL'").reset_index().date.map(lambda x: x.strftime("%m-%d-%Y")).to_dict()
df_post.reset_index(drop=True, inplace=True)
column_index = list(df_post.columns)
column_index.remove('sym')
column_index = dict(zip(range(len(column_index)), column_index))
all_symbols = df_post['sym'].unique()

In [78]:
symbol_index = tuple(all_symbols)

In [79]:
index = {
    'column': column_index,
    'date_range': date_range,
    'stocks': symbol_index
}
req = fb.refresh_index(index)